tests/pytest/test.py

```diff
@@ -1420,6 +1420,36 @@ def testRDBUnboundedDepth(env):
     r.expect('RESTORE', 'doc1', 0, dump).ok()
     r.expect('JSON.GET', 'doc1', '$..__leaf..__deep_leaf').equal('[420]')
 
+def test_promote_u64_to_f64(env):
+    r = env
+    i64max = 2 ** 63 - 1
+    u64max = 2 ** 64 - 1
+    
+    # i64 behaves normally
+    r.expect('JSON.SET', 'num', '$', i64max).ok()
+    res = r.execute_command('JSON.GET', 'num', '$')
+    val = json.loads(res)[0]
+    r.assertEqual(val, i64max)
+    r.assertNotEqual(val, float(i64max)) # i64max is not representable as f64
+
+    # i64 + i64 == u64 is promoted to f64
+    nn = random.randint(1, i64max)
+    res = r.execute_command('JSON.NUMINCRBY', 'num', '$', nn)
+    val = json.loads(res)[0]
+    r.assertNotEqual(val, -i64max + nn - 2)        # i64
+    r.assertNotEqual(val, i64max + nn)             # u64
+    r.assertEqual(val, float(i64max) + float(nn))  # f64
+
+    # set u64 is promoted to f64
+    nn = random.randint(i64max + 1, u64max)
+    while nn == float(nn):
+        nn = random.randint(i64max + 1, u64max)
+    r.expect('JSON.SET', 'num', '$', nn).ok()
+    res = r.execute_command('JSON.GET', 'num', '$')
+    val = json.loads(res)[0]
+    r.assertNotEqual(val, nn)     # u64
+    r.assertEqual(val, float(nn)) # f64
+
```

Comment:
Random values in unit tests cause them to be not repeatable. As soon as one test will pass and another will fail without any change, people lose confidence in them, undermining their value. Printing a reproduction script is not enough. That said, randomized edge case testing and fuzz testing can provide value.

Can we make the test here a little simpler, providing the correct `nn` value straight away?

Reply:
The point of the random values was that it should work for *every* value within the ranges provided. But looping over all 2^63 values in that range is a huge waste of time.

Reply:
This should be testable through the corner values: the min and close to min, somewhere in the middle, the max and close to the max. However, randomness in the tests may bring randomness in the test results. Indeed, looping isn't a good option.

Reply:
Randomness can be used and still allow for repeatable test execution, such as by logging the random seed, so a specific test failure can be reproduced. For example, https://github.com/RedisJSON/RedisJSON/blob/6ea15d7e795046f6170466ca232a790261718131/tests/pytest/test_short_read.py#L437